bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeConverter.java

```diff
@@ -1201,13 +1205,17 @@ private static boolean isFloatWithinIntRange(double doubleValue) {
     }
 
     public static Type resolveMatchingTypeForUnion(Object value, Type type) {
+        if (value instanceof TableValue && ((TableValue<?, ?>) value).getType().getTag() == TypeTags.TABLE_TAG) {
+            return ((TableValue<?, ?>) value).getType();
```

Comment:
Here just can't return the `getType()` because of the readonly status difference between source and target types, will fix and update and add tests